flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -53,6 +56,56 @@ public CheckpointFailureManager(int tolerableCpFailureNumber, FailJobCallback fa
         this.countedCheckpointIds = ConcurrentHashMap.newKeySet();
     }
 
+    /**
+     * Failures on JM:
+     *
+     * <ul>
+     *   <li>all checkpoints - go against failure counter.
+     *   <li>any savepoints - don’t do anything, manual action, the failover will not help anyway.
+     * </ul>
+     *
+     * <p>Failures on TM:
+     *
+     * <ul>
+     *   <li>all checkpoints - go against failure counter (failover might help and we want to notify
+     *       users).
+     *   <li>sync savepoints - we must always fail, otherwise we risk deadlock when the job
+     *       cancelation waiting for finishing savepoint which never happens.
+     *   <li>non sync savepoints - go against failure counter (failover might help solve the
+     *       problem).
+     * </ul>
+     *
+     * @param pendingCheckpoint the failed checkpoint if it was initialized already.
+     * @param checkpointProperties the checkpoint properties in order to determinate which handle
+     *     strategy can be used.
+     * @param exception the checkpoint exception.
+     * @param executionAttemptID the execution attempt id, as a safe guard.
+     */
+    public void handleCheckpointException(
+            @Nullable PendingCheckpoint pendingCheckpoint,
+            CheckpointProperties checkpointProperties,
+            CheckpointException exception,
+            @Nullable ExecutionAttemptID executionAttemptID) {
+        if (isJMFailure(pendingCheckpoint, exception, executionAttemptID)) {
+            handleJobLevelCheckpointException(
+                    checkpointProperties,
+                    exception,
+                    pendingCheckpoint == null
+                            ? UNKNOWN_CHECKPOINT_ID
+                            : pendingCheckpoint.getCheckpointID());
+        } else {
+            handleTaskLevelCheckpointException(pendingCheckpoint, exception, executionAttemptID);
+        }
+    }
+
+    /** Check if the exception occurs on the job manager side or not. */
+    private boolean isJMFailure(
+            @Nullable PendingCheckpoint pendingCheckpoint,
+            CheckpointException exception,
+            @Nullable ExecutionAttemptID executionAttemptID) {
+        return pendingCheckpoint == null || isJMThrowable(exception) || executionAttemptID == null;
```

Comment:
Hmmm, since we have `CheckpointFailureReason#isPreFlight`, that begs a question, why do we even need the other conditions? 

About `pendingCheckpoint == null` (that's the new feature), I would actually try to get rid of it and make sure that `CheckpointFailureReason#isPreFlight` returns true for every case when we could expect `pendingCheckpoint == null` (go through all of the paths leading to it)? I'm not sure how big of an effort that would be.

About `executionAttemptID == null;` I would do the same, but that's more risky, as it can cause a regression very late in the release process, so I would keep it as it is, with a `TODO` for the future. And maybe immediately after fixing this bug follow up on it, and try to get rid of it as well?

Reply:
I agree that CheckpointFailureReason#isPreFlight should be enough but it was just extra protection. 
I removed `pendingCheckpoint == null` because it indeed can rely on isPreFlight method but `executionAttemptID == null` is more complicated so I left it as you proposed.